Remove deprecation warnings to prepare for Gradle 5 (sourceSets.main.output.classesDirs)

alpar-t opened the pull request:
Gradle replaced `project.sourceSets.main.output.classesDir` of type
`File` with `project.sourceSets.main.output.classesDirs` of type
`FileCollection`
(see [SourceSetOutput](https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java))
Build output is now stored on a per language folder.

There are a few places where we use that, here's these and how it's
fixed:

- Randomized Test execution
    - look in all test folders ( pass the multi dir configuration to the
    ant runner )
    - DRY the task configuration by introducing `basedOn` for
      `RandomizedTestingTask` DSL
- Extend the naming convention test to support passing in multiple
  directories
- Fix the standalon test plugin, the dires were not passed trough,
  checked with a debuger and the statement had no affect due to a
  missing `=`.

Closes  #30354

alpar-t committed:
Remove deprecation warnings to prepare for Gradle 5

Gradle replaced `project.sourceSets.main.output.classesDir` of type
`File` with `project.sourceSets.main.output.classesDirs` of type
`FileCollection`
(see [SourceSetOutput](https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java))
Build output is now stored on a per language folder.

There are a few places where we use that, here's these and how it's
fixed:

- Randomized Test execution
    - look in all test folders ( pass the multi dir configuration to the
    ant runner )
    - DRY the task configuration by introducing `basedOn` for
      `RandomizedTestingTask` DSL
- Extend the naming convention test to support passing in multiple
  directories
- Fix the standalon test plugin, the dires were not passed trough,
  checked with a debuger and the statement had no affect due to a
  missing `=`.

Closes #30354

dadoonet commented:
Related to #27250.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy:
```diff
@@ -81,11 +80,18 @@ public class NamingConventionsTask extends LoggedExec {
+
+        def dirsToUse = (checkForTestsInMain ?
+            project.sourceSets.main.output.classesDirs :
+            project.sourceSets.test.output.classesDirs)
```
I don't think we can check these here, this would still be during configuration? Certainly not the existence of them in the filter below.

atorok replied:
This works because the FileCollection is a live collection, this the checks are actually only executed lather.
I'm changing this anyhow, as  I discovered that we should not pick up non Java tests anyhow, because those don't adhere to the naming conventions.

rjernst commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy:
```diff
@@ -95,6 +97,17 @@ class RandomizedTestingTask extends DefaultTask {
     }
 
+    void basedOn(RandomizedTestingTask other) {
+        configure(BuildPlugin.commonTestConfig(project))
```
BuildPlugin should not be referred to here. The intention has always been to move the randomized testing gradle code (notice this is in a different java package than the rest of the buildSrc code) back upstream with randomized testing itself. See https://github.com/randomizedtesting/gradle-randomized-testing-plugin where I made some snapshots when first working on migrating the build to gradle, but was not able to cleanup and document it enough to release on its own.

atorok replied:
I  moved it back to the Gradle script.

rjernst commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy:
```diff
@@ -95,6 +97,17 @@ class RandomizedTestingTask extends DefaultTask {
+    void basedOn(RandomizedTestingTask other) {
+        configure(BuildPlugin.commonTestConfig(project))
+        classpath = other.classpath;
+        testClassesDirs = other.testClassesDirs;
```
I think all the users of this could just be converted to use this line? I don't think we need to or should overly abstract this "shared code". For example, it would be better to make internalClusterTest task added by an additional gradle plugin to reduce duplication there.

atorok replied:
I did change just the line everywhere during development, and decided to take the extra step to refactor it as I felt it's too detailed and does not convey intent well enough, every time I encountered it I had to carefully check to make sure none of the configuration is missing or slightly different. I do agree that a plugin as you suggest would be even better,  but this is a step that will make that a bit easier to implement.

rjernst reviewed (changes requested):
Thanks for starting on this! I've left some comments. More generally, I see a lot of unrelated changes here. Things like reformatting or restructuring should generally go in separate PRs to ease reviewing.

alpar-t committed:
Merge branch 'master' into fix/30354

atorok commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingPlugin.groovy:
```diff
@@ -44,24 +44,24 @@ class RandomizedTestingPlugin implements Plugin<Project> {
 
-    static void replaceTestTask(TaskContainer tasks) {
+    static void replaceTestTask(Project project) {
+        def tasks = project.tasks
```
fixed

atorok commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy:
```diff
@@ -211,7 +224,7 @@ class RandomizedTestingTask extends DefaultTask {
         DefaultLogger listener = null
         ByteArrayOutputStream antLoggingBuffer = null
-        if (logger.isInfoEnabled() == false) {
+        if (!logger.isInfoEnabled()) {
```
fixed

alpar-t committed:
Only check Java tests, PR feedback

- Name checker was ran for Groovy tests that don't adhere to the same
  convections causing the check to fail
- implement PR feedback

alpar-t committed:
Merge branch 'master' into fix/30354

rjernst commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy:
```diff
@@ -211,7 +223,7 @@ class RandomizedTestingTask extends DefaultTask {
         DefaultLogger listener = null
         ByteArrayOutputStream antLoggingBuffer = null
-        if (logger.isInfoEnabled() == false) {
+        if (false == logger.isInfoEnabled()) {
```
I still see lots of arbitrary syntactical changes like this. Every extra line of output makes something that must be looked at. Without these minor changes, I think this PR would only be a dozen lines or so and would be very quick to review.

atorok replied:
I don't have a problem omitting them if that's the consensus, my take on it is that the cost of making and reviewing these is worthwhile if they improve the readability of the code, as it eventually pays off.

rjernst commented on buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy:
```diff
@@ -95,6 +97,16 @@ class RandomizedTestingTask extends DefaultTask {
         listenersConfig.listeners.add(new TestReportLogger(logger: logger, config: testLoggingConfig))
     }
 
+    void basedOn(RandomizedTestingTask other) {
```
I really would like to not have this. It makes many assumptions (eg that that `check` should depend on this). Can we please keep this change only to updating testClassesDirs and look at ways to deduplicate internClusterTest in a separate issue?

atorok replied:
Removed it. I have a better understanding of why this was not a good idea now, and why this would be better handled in it's own PR.  There are multiple ways to manage tests, deal with unit and integrations tests etc ( there's an `additionalTests` extension that is similar to what `basedOn` did ).  It might also be worth setting up a different source path for integration tests then we could auto wire that to check and offer a clear DSL for defining new tests that are possibly  a subset of the existing unit or integration with additional system properties.

rjernst reviewed (commented):
I left a few more comments

alpar-t committed:
Replace `add` with `addAll`

This worked because the list is passed to `project.files` that does the
right thing.

alpar-t committed:
Revert "Only check Java tests, PR feedback"

This reverts commit 9bd9389875d8b88aadb50df57a45cd0d2b073241.

alpar-t committed:
Remove `basedOn` helper

alpar-t committed:
Bring some changes back

Previus revert accidentally reverted too much

alpar-t committed:
Fix negation

atorok commented:
build this please

atorok commented:
test this please

atorok commented:
test this please

alpar-t committed:
Merge commit '9434f25' into fix/30354

atorok commented:
test this please

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy:
```diff
@@ -66,7 +65,7 @@ public class NamingConventionsTask extends LoggedExec {
     boolean checkForTestsInMain = false;
 
-    public NamingConventionsTask() {
+    NamingConventionsTask() {
```
Please add back the `public`. As I mentioned before, a goal here is to keep these files as close as possible to java so buildSrc can be converted completely to java (hopefully someday soon).

rjernst replied:
Kotlin was not around when we switched to gradle. We have tried sticking to java for multiple reasons:
* Expertise on Elasticsearch is usually in java
* I have seen numerous reports of double digit performance gains in configuration time from switching gradle plugins to java

Remember we are only talking about buildSrc here. The build.gradle files are still groovy. I am not averse to them being Kotlin, but I would want to see a significant advantage to that work, as we have now spent over 2 years with developers learning how to write groovy in gradle DSL.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/test/NamingConventionsCheck.java:
```diff
@@ -70,18 +72,21 @@ public static void main(String[] args) throws IOException {
-            check.checkMain(rootPath);
-        } else {
-            check.checkTests(rootPath, skipIntegTestsInDisguise);
+        for (String rootDir : rootPathList.split(Pattern.quote(File.pathSeparator))) {
```
This seems inconsistent with the gradle code. There, a single dir is currently passed in. But here it looks like the path of that dir is split into its elements but then each read as a dir...

rjernst replied:
I don't think we should do that. Each project should be checkable on its own. Combining all the sources into one task means to run precommit on a project requires running this one large task that looks at all test classes.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy:
```diff
@@ -111,16 +111,27 @@ public class NamingConventionsTask extends LoggedExec {
+        // see https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSetOutput.html the deprecated property
+        // classesDir is the first in the FileCollection. There seems to be no other way to refer to the Java output dir
+        // We can't run the check against all classes as Groovy tests don't adhere to naming conventions.
+        return classesDirs[0]
```
Why don't we change the java code to work on multiple directories? Then you can just pass in all the classesDirs as multiple arguments.

rjernst replied:
Where do we have have groovy tests? How do we know the ordering would always result in the java classes dir coming first?

atorok replied:
we have groovy  `./buildSrc/src/test/groovy/`. It's an odd that the Gradle API does not offer a more explicit way to get the java specific dirs, and choose to document that it will always be first here: 
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSetOutput.html 
It mentions the following for the deprectaed `File classesDir`  property:
```
Default with java plugin:
First directory from classesDirs
```

rjernst replied:
While that deprecated property may document that fact, there would be nothing preventing gradle from changing the order once that property is removed. There are only 2 groovy tests. I think we can update their names to follow the naming convention?

atorok replied:
@rjernst I updated the task to pass along all dirs, but also added a new input to be able to bypass that and be specific of the  directory. Adding the Groovy tests is trickier because of the inheritance check but we'll be ok, since the build tools that have a hard coded path now are self tests, so if the test would change the test would fail.

alpar-t committed:
add back public

alpar-t committed:
Merge commit 'c6be3b4' into fix/30354

alpar-t committed:
revert name check changes

alpar-t committed:
Revert "revert name check changes"

This reverts commit a2800c0b363168339ea65e2a79ec8256e5883e6d.

alpar-t committed:
Pass all dirs to name check

Only run on Java for build-tools, this is safe because it's a self test.
It needs more work before we could pass in the Groovy classes as well as
these inherit from `GroovyTestCase`

alpar-t committed:
Merge commit 'e888467d0a7af4115789d08e4cd96bb98375968e' into fix/30354

alpar-t committed:
remove self tests from name check

The self complicates the task setup and disable real checks on
build-tools.
With this change there are no more self tests, and the build-tools tests
adhere to the conventions.
The self test will be replaced by gradle test kit, thus the addition of
the Gradle plugin builder plugin.

alpar-t committed:
First test to run a Gradle build

alpar-t committed:
Add tests that replace the name check self test

atorok commented:
I wasn't really happy about this solution as I realized that it makes it hard to add java tests to the `buildSrc` which would be a big no-no. I decided to get rid of the self test in favor of a proper integration test with gradles testkit. As part of the change I also converted some files to Java using automated IDE tools to better prove the point of java test - java class and updated the Groovy tests to

alpar-t committed:
Merge commit '8e4ab82' into fix/30354

atorok commented:
test this please

atorok commented:
test this please

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionProperties.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Accessor for shared dependency versions used by elasticsearch, namely the elasticsearch and lucene versions.
+ */
+public class VersionProperties {
```
It appears this is being moved to a java file, but it is still under `src/main/groovy`?

atorok replied:
Yes. That is because the Groovy compiler knows how to deal with this and delegate these files to the Java compiler. The reason these can't move to `src/main/java` (yet) us that they make use of Groovy classes that are not on the cp  when pure java is compiled. Moving those Groovy classes to Java creates a ripple effect, so my thinking is that we can have java files in the groovy source set until we have critical mass, eventually removing the groovy source set. It's easier and less risky than converting to Java at once.

rjernst replied:
This sounds like a great plan!

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/LoggedExec.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * A wrapper around gradle's Exec task to capture output and log on error.
+ */
+public class LoggedExec extends Exec {
```
This appears to be moved to a java file, but still exists under `src/main/groovy`?

rjernst commented on buildSrc/src/main/java/org/elasticsearch/test/NamingConventionsCheck.java:
```diff
@@ -63,59 +59,61 @@ public static void main(String[] args) throws IOException {
+                    check.integTestsInDisguise
+                );
         }
+        System.exit(exitCode);
```
I don't think we should call System.exit. Instead, throw a GradleException with a message indicating there were violations.

atorok replied:
This uses an exit code since we don't call into this code directly, but run it like a cli utility.

rjernst commented on buildSrc/src/test/groovy/org/elasticsearch/gradle/VersionCollectionTests.groovy:
```diff
@@ -16,6 +20,7 @@ class VersionCollectionTest extends GroovyTestCase {
    * branched from Major-1.x At the time of this writing 6.2 is unreleased and 6.3 is the 6.x branch. This test simulates the behavior
    * from 7.0 perspective, or master at the time of this writing.
    */
+  @Test
```
I think this will be confusing to start having @Test in gradle tests, when we do not have them in the rest of the tests (and it is unclear how this would work when we run this as a normal `elasticsearch.build` as the `build-tools` project that uses randomized runner). We should be able to annotate the test with a junit3 compatible selector like randomized runner does?

atorok replied:
We could inherit from TestCase but we loose `@Rule` and probably `@Before`, `@After` etc. 
The rest of the code supports both `@Test` and methods starting with `test`, but this is implemented in `@RunWith(RandomizedRunner.class)`, `@TestMethodProviders` is _not_ a junit4 annotation, the junit3 compatible selector is part of lucene test, so while it's possible to bring that in as a dependency it seems a bit odd. The other possibility is to duplicate that one filter class.

rjernst replied:
I would duplicate the class if necessary. Having @Test in some parts of our code would be very confusing.

atorok replied:
done

rjernst commented on buildSrc/src/test/groovy/org/elasticsearch/gradle/doc/RestTestsFromSnippetsTaskTest.groovy:
```diff
@@ -19,32 +19,46 @@
+import org.junit.Test
+import org.junit.rules.ExpectedException
+
+import static org.junit.Assert.*
```
Please do not use wildcard imports. We disallowed them in the rest of the codebase.

atorok replied:
I fixed my IDE setting some time ago, this must pre-date that. Fixed now.

rjernst reviewed (commented):
This is looking closer, but I have a few more comments.

alpar-t committed:
Merge commit '8b698f0' into fix/30354

alpar-t committed:
Clean up integ test base class

alpar-t committed:
Always run tests

alpar-t committed:
Align with test naming conventions

alpar-t committed:
Make integ. test case inherit from unit test case

The check requires this

rjernst commented on buildSrc/src/test/groovy/org/elasticsearch/gradle/VersionCollectionTests.groovy:
```diff
@@ -1,6 +1,11 @@
+import org.elasticsearch.gradle.test.GradleUnitTestCase
+import org.junit.Test
+
+import static org.junit.Assert.*
```
Still have a wildcard import

rjernst commented on buildSrc/src/test/groovy/org/elasticsearch/gradle/doc/RestTestFromSnippetsTaskTests.groovy:
```diff
@@ -19,31 +19,42 @@
+
 import static org.elasticsearch.gradle.doc.RestTestsFromSnippetsTask.replaceBlockQuote
+import static org.elasticsearch.gradle.doc.RestTestsFromSnippetsTask.shouldAddShardFailureCheck
+import static org.junit.Assert.*
```
Still have a wildcard import

rjernst commented on buildSrc/src/test/java/org/elasticsearch/gradle/test/BaseTestCase.java:
```diff
@@ -0,0 +1,32 @@
+    JUnit4MethodProvider.class,
+    JUnit3MethodProvider.class
+})
+public abstract class BaseTestCase {
```
You can make this extend from `Assert` to not have to import the common methods used like assertEquals

atorok replied:
I'll fix these, thanks, I finally figured out  where these are coming from, it's in the template for test classes.

rjernst commented on buildSrc/src/test/java/org/elasticsearch/gradle/test/GradleUnitTestCase.java:
```diff
@@ -0,0 +1,14 @@
+    JUnit4MethodProvider.class,
+    JUnit3MethodProvider.class
+})
+public abstract class GradleUnitTestCase extends BaseTestCase  {
```
Why have this and BaseTestCase? I think just having one of these would be fine?

atorok replied:
An integration test is not a unit test, so I didn't initially want them to inherit from each-other, but would have to extend the name checker to be able to deal with it.

rjernst reviewed (approved):
LGTM, a few last comments.

alpar-t committed:
Remove `import static org.junit.Assert.*`

alpar-t committed:
Merge branch 'master' into fix/30354
